kstore-file/src/commonTest/kotlin/io/github/xxfast/kstore/file/FileCodecTests.kt

```diff
@@ -37,23 +38,35 @@ class FileCodecTests {
 
   @Test
   fun testEncode() = runTest {
-    codec.encode(MYLO)
-    val expect: Cat = MYLO
-    val actual: Cat? = stored
+    codec.encode(listOf(MYLO))
+    val expect: List<Pet> = listOf(MYLO)
+    val actual: List<Pet>? = stored
     assertEquals(expect, actual)
   }
 
   @Test
   fun testDecode() = runTest {
-    stored = OREO
-    val expect: Cat = OREO
-    val actual: Cat? = codec.decode()
+    stored = listOf(OREO)
+    val expect: List<Pet> = listOf(OREO)
+    val actual: List<Pet>? = codec.decode()
     assertEquals(expect, actual)
   }
 
+  @Test
+  fun testTransactionalEncode() = runTest {
+    codec.encode(listOf(MYLO))
+
+    // Encoder will fail half way through
+    assertFailsWith<NotImplementedError> { codec.encode(listOf(MYLO, KAT)) }
+
+    // The original file should not be touched
+    val actual: List<Pet>? = codec.decode()
+    assertEquals(listOf(MYLO), actual)
```

Comment:
Nice work @xxfast! I’m wondering what happens if a race condition occurs and two different agents/threads start writing to a file at the same time, but one of them gets canceled or stopped. Does this revert the file back to an older version that was not modified? What happens to the version of the file being updated by the other agent/thread?

I suggest there should be a way to check if someone is modifying the file, and the other process should have to wait or be queued until the current modification is finished.

Reply:
There should really be a single store for a single file, but if there's an external process that is writing to the same file - and if the store write was cancelled, it should delete the temp file and leave the original file untouched